lib/Backends/NNPI/InferenceContext.cpp

```diff
@@ -0,0 +1,433 @@
+  }
+  LOG_NNPI_INF_IF_ERROR_RETURN_FALSE(
+      nnpiInferCommandCreate(deviceNetwork, inputHandles, numInputs,
+                             outputHandles, numOutputs, &inferCmd_),
+      "Failed to create NNPI inference command");
+
+  if (deviceOptions_->enabledCommandLists > 0) {
+    // collect copy commands for the list (some resources may not need copying).
+    std::vector<NNPICommandHandle> commands;
+    std::vector<NNPICopyCommand> inputCopyCmds, outputCopyCmds;
+    for (auto &res : inputResources_) {
+      auto copyCmd = res->GetCopyCommand();
+      if (copyCmd) {
+        res->SetCmdListIdx(static_cast<uint32_t>(commands.size()));
+        NNPICommandHandle cmd;
+        cmd.type = NNPI_COMMAND_TYPE_COPY;
+        cmd.copyCommand = copyCmd;
+        commands.push_back(cmd);
+      }
+    }
+    {
+      NNPICommandHandle cmd;
+      cmd.type = NNPI_COMMAND_TYPE_INFER;
+      cmd.inferCommand = inferCmd_;
+      commands.push_back(cmd);
+    }
+    for (auto &res : outputResources_) {
+      auto copyCmd = res->GetCopyCommand();
+      if (copyCmd) {
+        res->SetCmdListIdx(static_cast<uint32_t>(commands.size()));
+        NNPICommandHandle cmd;
+        cmd.type = NNPI_COMMAND_TYPE_COPY;
+        cmd.copyCommand = copyCmd;
+        commands.push_back(cmd);
+      }
+    }
+
+    // Create command list.
+    LOG_NNPI_INF_IF_ERROR_RETURN_FALSE(
+        nnpiCommandListCreate(&(commands[0]),
+                              static_cast<uint32_t>(commands.size()), nullptr,
+                              0, &commandList_),
+        "Failed to create NNPI command list");
+
+    // Preallocate enough configs to be used for partials later on.
+    cmdConfigs_.resize(commands.size());
+    // Preallocate enough errors to be used later durin inference.
+    cmdListErrors_.resize(commands.size());
+  }
+
+  return true;
+}
+
+void InferenceContext::execute(RunIdentifierTy runId,
+                               std::unique_ptr<ExecutionContext> ctx,
+                               runtime::ResultCBTy resultCB) {
+  TRACE_EVENT_SCOPE(ctx->getTraceContext(), TraceLevel::REQUEST,
+                    TRACING_BACKEND_EXECUTE);
+  if (ctx->getTraceContext()) {
+    ctx->getTraceContext()->setThreadName("InferenceContext");
+  }
+
+  // Pre inference input preparation.
+  PlaceholderBindings &bindings = *ctx->getPlaceholderBindings();
+
+  std::unordered_set<Tensor *> partialTensorInputs;
+  for (auto &pht : bindings.pairs()) {
+    if (partialInputs_->count(pht.first)) {
+      partialTensorInputs.insert(pht.second);
+    }
+  }
+  TRACE_EVENT_BEGIN(ctx->getTraceContext(), TraceLevel::COPY,
+                    TRACING_PRE_PROCESS);
+
+  // Pre-inference
+  auto &phTensorMap = bindings.pairs();
+  std::vector<void *> rawInputs, rawOutputs;
+  for (auto &in : inputResources_) {
+    if (in->GetUsage() != NNPIResource::ResourceUsage::StaticInputResource) {
+      Placeholder *ph = bindings.getPlaceholderByName(in->GetName());
```

Comment:
This seems to be a regression from recent positional bindings as here we are doing string lookup per inference per input again.

Reply:
Yes, we can apply a similar change as in #4147 here.

Reply:
Should we do it in this PR or a follow-up?

Reply:
Will try to do that on Sunday - otherwise can do it later in the week.